### PR TITLE
added multiprocessed, added support for converting masks, depth and n…

### PR DESCRIPTION
**Pull request by Shreeyak**
…ormals (1-channel, uint16, float16 or float32)

### CODE REVIEW

**Comment by Shreeyak** [comment]
@adavydow For masks, depth and normals, we want to have nearest-neighbor interpolation during distortion. Could you add that fix?

**Comment by Shreeyak** [comment]
Updated requirements and readme.

File: README.md [hunk 1]
```diff
@@ -31,11 +33,20 @@ This will process all the images found in the container's `/data` directory. Mod
...
-dir_output: images/
-input_rgb_ext: .rgb.png  # Distortion will be applied to all images matching this extension
-input_json_ext: .json  # The camera intrinsics are loaded from this json file
+dir_input: samples/
+dir_output: samples/
+input:
+  # Distortion will be applied to these files (except info).
+  # Set their name to null to disable distorting those images
+  rgb: .rgb.png
+  segments: .segments.png  # Mask of all objects in scene. 8 or 16 bit PNG.
```
Comment (adavydow) [inline README.md:44]: I'd prefer to infer # of channels and image type from the image itself and provide interpolation type as additional parameter.
I'm not sure if hardcoding scene type is reasonable here.

File: apply_fisheye_distortion.py [hunk 2]
```diff
@@ -64,10 +102,100 @@ def distort_image(img: np.ndarray, cam_intr: np.ndarray, dist_coeff: np.ndarray,
...
+    class FileMode:
+        path: Union[Path, None]
+        mode: DistortMode
+
+    _f_rgb = FileMode(f_rgb, DistortMode.LINEAR)
+    _f_segments = FileMode(f_segments, DistortMode.NEAREST)
+
+    list_files = [_f_rgb, _f_segments]
+    for _f in list_files:
+        _process_file(f_json=f_json, f_img=_f.path, dir_output=dir_output, dist_coeff=dist_coeff, mode=_f.mode,
```
Comment (adavydow) [inline apply_fisheye_distortion.py:170]: Here None is allowed for f_rgb and f_segments, but _process_file does not accept None as f_img path.

File: apply_fisheye_distortion.py [hunk 3]
```diff
@@ -77,58 +205,52 @@ def main():
...
+    # process_files(f_json: Path, f_rgb: Union[Path, None], f_segments: Union[Path, None], dir_output: Path,
+    # dist_coeff: np.ndarray, crop_resize_output: bool)
+
+    if int(cfg.workers) > 0:
+        max_workers = int(cfg.workers)
+    else:
+        max_workers = None
+    with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
+        with tqdm(total=len(json_filenames)) as pbar:
+            for _ in executor.map(process_files, json_filenames, rgb_filenames, segments_filenames,
```
Comment (adavydow) [inline apply_fisheye_distortion.py:250]: Looks like this will not work if number of segments and rgb filenames mismatch.

Comment (adavydow) [reply apply_fisheye_distortion.py:250]: Oh, ok. It will fail in check_num images then. Any reason check_num_images returns number of images as second element of the pair? It should always be equal to num_json, right?

File: apply_fisheye_distortion.py [hunk 4]
```diff
@@ -64,10 +102,100 @@ def distort_image(img: np.ndarray, cam_intr: np.ndarray, dist_coeff: np.ndarray,
...
+                             f'  dir: "{dir_input}"\n'
+                             f'  file extention: "{input_file_ext}"')
+        elif num_images != num_json:
+            raise ValueError(f'Unequal number of json files and images:\n'
+                             f'  num json: {num_json}\n'
+                             f'  num images: {num_images}\n'
+                             f'  dir: "{dir_input}"\n'
+                             f'  file extention: "{input_file_ext}"')
+
+    return image_filenames, num_images
```
Comment (adavydow) [inline apply_fisheye_distortion.py:194]: Any reason to return num_images?

Comment (adavydow) [reply apply_fisheye_distortion.py:194]: Got it, it can be 0.
Logic here is really hard to understand and follow.

File: apply_fisheye_distortion.py [hunk 5]
```diff
@@ -64,10 +102,100 @@ def distort_image(img: np.ndarray, cam_intr: np.ndarray, dist_coeff: np.ndarray,
...
+    else:
+        retval = cv2.imwrite(str(out_filename), dist_img)
+        if retval:
+            pass
+            # print(f'exported image: {out_filename}')
+        else:
+            raise RuntimeError(f'Error in saving file {out_filename}')
+
+
+def process_files(f_json: Path, f_rgb: Union[Path, None], f_segments: Union[Path, None], dir_output: Path,
```
Comment (adavydow) [inline apply_fisheye_distortion.py:149]: Is this function needed? I think it'd be easier to just call process_file twice in the main loop.

File: apply_fisheye_distortion.py [hunk 6]
```diff
@@ -45,10 +70,23 @@ def distort_image(img: np.ndarray, cam_intr: np.ndarray, dist_coeff: np.ndarray,
...
+        img = np.expand_dims(img, 2)
+    interpolators = [scipy.interpolate.RegularGridInterpolator((ys, xs), img[:, :, channel], method=mode.value,
+                                                               bounds_error=False, fill_value=0)
+                     for channel in range(chan)]
     img_dist = np.dstack([interpolator(undistorted_px) for interpolator in interpolators])
-    img_dist = img_dist.clip(0, 255).astype(np.uint8)
+
+    if imdtype == np.uint8:
+        # RGB Image
+        img_dist = img_dist.clip(0, 255).astype(np.uint8)
```
Comment (adavydow) [inline apply_fisheye_distortion.py:82]: I'd prefer explicit rounding. Also, currently `0.9` is treated as `0` and you will almost never have pixels with value 255.

**Review by adavydow — CHANGES_REQUESTED** [review]
Major: the way to parse None for rgb or segment index is currently extremely overcomplicated and hard to follow.
Minor: do we need to hardcode scenetypes? Isn't passing only interpolation method and extension mask a more versatile solution?

**Comment by Shreeyak** [comment]
> Major: the way to parse None for rgb or segment index is currently extremely overcomplicated and hard to follow.
> Minor: do we need to hardcode scenetypes? Isn't passing only interpolation method and extension mask a more versatile solution?

Both of these issues are intertwined. The aim of this repo is that we can process multiple types of files in one go (rgb, segments, depth, normals) instead of calling the script multiple times. I also don't want to give the interpolation as a user input since it will result in user errors.
The config file should be easily modified such that we only process the type of file we want. I'll rework the logic to make it easier to follow, and I'll also rework the config so it's easier to use as intended:

```yaml
dir_input: samples/
dir_output: samples/
workers: 0  # How many workers to use to convert files. 0 for all cores.
crop_and_resize_output: True  # Disable to get original distorted image
apply_dist:
  rgb: True
  segments: True

distortion_parameters:
  # Ref: https://docs.opencv.org/4.4.0/db/d58/group__calib3d__fisheye.html
  # These are the distortion parameters of the fisheye camera model as defined in the fisheye module of OpenCV 4.4.0
  k1: 0.17149
  k2: -0.27191
  k3: 0.25787
  k4: -0.08054

file_ext:
  rgb: .rgb.png
  segments: .segments.png  # Mask of all objects in scene. 8 or 16 bit PNG.
  info: .json  # Contains camera intrinsics
```

File: apply_fisheye_distortion.py (hunk 5 above)
Comment (Shreeyak) [reply apply_fisheye_distortion.py:149]: This part needs reworking. I want to process all types of files belonging to the same render ID before proceeding to the next render ID. 
I'll prob keep track of render ID directly and append the file extensions to get the various files.

File: README.md (hunk 1 above)
Comment (Shreeyak) [reply README.md:44]: This allows users to specify which types of files they want to convert. And allows multiple types of files to be converted at once, instead of calling the script multiple times.
While your approach would work better to accomodate for any file extension (eg .distorted.rgb.png), it puts the onus on the user to figure out the file extension for each type of file they want to convert. I prefer to code it into the config, since we have a few standard extensions (optionally, config value can be overridden).

What do you feel is better?

Comment (adavydow) [reply README.md:44]: Personally I'd prefer to leave it for user, but I think both ways are ok.

**Comment by adavydow** [comment]
I don't see how are this issues intertwined. The second one is not a big problem and I'm ok with it.
If you want to leave config as is, please process None's explicitly in the code. Current way of dealing with it is extremely overcomplicated and hard to understand/support.

**Comment by Shreeyak** [comment]
After using the repo, it's clear that your suggested method is more flexible. I've updated the code according to your suggestions (user needs to supply the filename extention and the type of interpolation). This also simplifies the code and clears the issue you raised about dealing with "None's".

Since this script is needed now for some deliverables, I'll go ahead and merge the PR.

File: README.md (hunk 1 above)
Comment (Shreeyak) [reply README.md]: Pushed fix as suggested (leave it for user)

File: apply_fisheye_distortion.py (hunk 2 above)
Comment (Shreeyak) [reply apply_fisheye_distortion.py]: Sorted in new fix

File: apply_fisheye_distortion.py (hunk 4 above)
Comment (Shreeyak) [reply apply_fisheye_distortion.py]: All this logic removed in new fix, since user specifies a single file ext to process.
